models/VQVAE.py

```diff
@@ -0,0 +1,258 @@
+        pre_vq_x = self.pre_vq_conv_layer(encoded_x)
+        (
+            quantized_x,
+            dictionary_loss,
+            commitment_loss,
+            encoding_indices,
+        ) = self.vector_quantizer(pre_vq_x)
+        x_recon = self.decoder_layer(quantized_x)
+
+        return {
+            "dictionary_loss": dictionary_loss,
+            "commitment_loss": commitment_loss,
+            "x_recon": x_recon,
+        }
+
+
+class SonnetExponentialMovingAverage(nn.Module):
+    # See: https://github.com/deepmind/sonnet/blob/5cbfdc356962d9b6198d5b63f0826a80acfdf35b/sonnet/src/moving_averages.py#L25.
+    # They do *not* use the exponential moving average updates described in Appendix A.1
+    # of "Neural Discrete Representation Learning".
+    def __init__(self, decay, shape):
+        super().__init__()
+        self.decay = decay
+        self.counter = 0
+        self.register_buffer("hidden", torch.zeros(*shape))
+        self.register_buffer("average", torch.zeros(*shape))
+
+    def update(self, value):
+        self.counter += 1
+        with torch.no_grad():
+            self.hidden -= (self.hidden - value) * (1 - self.decay)
+            self.average = self.hidden / (1 - self.decay**self.counter)
+
+    def __call__(self, value):
+        self.update(value)
+        return self.average
+
+
+class VectorQuantizer(nn.Module):
+    def __init__(self, embedding_dim, num_embeddings, use_ema, decay, epsilon):
+        super().__init__()
+        # See Section 3 of "Neural Discrete Representation Learning" and:
+        # https://github.com/deepmind/sonnet/blob/v2/sonnet/src/nets/vqvae.py#L142.
+
+        self.embedding_dim = embedding_dim
+        self.num_embeddings = num_embeddings
+        self.use_ema = use_ema
+        # Weight for the exponential moving average.
+        self.decay = decay
+        # Small constant to avoid numerical instability in embedding updates.
+        self.epsilon = epsilon
+
+        # Dictionary embeddings.
+        limit = 3**0.5
+        e_i_ts = torch.FloatTensor(embedding_dim, num_embeddings).uniform_(-limit, limit)
+        if use_ema:
+            self.register_buffer("e_i_ts", e_i_ts)
+        else:
+            self.register_parameter("e_i_ts", nn.Parameter(e_i_ts))
+
+        # Exponential moving average of the cluster counts.
+        self.N_i_ts = SonnetExponentialMovingAverage(decay, (num_embeddings,))
+        # Exponential moving average of the embeddings.
+        self.m_i_ts = SonnetExponentialMovingAverage(decay, e_i_ts.shape)
+
+    def forward(self, x):
+        flat_x = x.permute(0, 2, 3, 1).reshape(-1, self.embedding_dim)
+        distances = (flat_x**2).sum(1, keepdim=True) - 2 * flat_x @ self.e_i_ts + (self.e_i_ts**2).sum(0, keepdim=True)
+        encoding_indices = distances.argmin(1)
+        quantized_x = F.embedding(encoding_indices.view(x.shape[0], *x.shape[2:]), self.e_i_ts.transpose(0, 1)).permute(
+            0, 3, 1, 2
+        )
```

Comment:
You could try to use einops package for pytorch: https://einops.rocks, it makes your life easier when you are reshaping, permuting and transposing tensors and also your code is easier to read :)

Reply:
For example:
```python
flat_x = x.permute(0, 2, 3, 1).reshape(-1, self.embedding_dim)
```
changes to
```python
flat_x = einops.rearrange(x, "b c h w -> (b h w) c")
```